### PR TITLE
Add authorization rules branch

### PR DESCRIPTION
**Pull request by kayjeee**
# Pull Request: Adding CanCanCan for Authorization

## Description

This pull request introduces CanCanCan to enhance the authorization system in our project. The key features added include the installation of CanCanCan, the addition of a role column to the users table, and the implementation of post and comment deletion functionality based on user roles.

## Changes Made

- Installed and configured CanCanCan in the project to improve authorization.

### Role-Based Authorization

- Added a role column to the users table using a migration.

#### Post Deletion Authorization

- Users can now delete a post if it belongs to them or if they have an admin role (role column with a value "admin"). CanCanCan is used to enforce this authorization.
- Implemented the post deletion functionality, including adding a "Delete" button to the view, ensuring that only authorized users can see and use it.

#### Comment Deletion Authorization

- Users can delete a comment if it belongs to them or if they have an admin role (role column with a value "admin"). CanCanCan is used to enforce this authorization.
- Implemented the comment deletion functionality, including adding a "Delete" button to the view, ensuring that only authorized users can see and use it.

## Checklist

- [x] Installed CanCanCan for role-based authorization.
- [x] Added a role column to the users table using migration.
- [x] Implemented post deletion functionality with CanCanCan authorization.
- [x] Added a "Delete" button for posts, visible only to authorized users.
- [x] Implemented comment deletion functionality with CanCanCan authorization.
- [x] Added a "Delete" button for comments, visible only to authorized users.

I am working solo on this project, as confirmed by this screenshot:
Solo Work

Please review and merge this pull request to incorporate the CanCanCan authorization system and the added features for post and comment deletion based on user roles.

### CODE REVIEW

**Review by nedjwab — APPROVED** [review]
# 🟢 Project Approved! 🏆

## Hi  @kayjeee                     👋

Your project is complete! ✔️✔  🏁

Your code was very professional 💯

There is nothing else to say other than... _it's time to merge it_ :shipit:
Congratulations! 🎉

 
## Highlights 🌟

- ✔️ Descriptive commmit messsages
- ✔️ Descriptive PR description
- ✔️ No linter Errors


- The code is top-notch! 💯 and have a huge potential.

Cheers and Happy coding!👏👏👏

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.


_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by kayjeee** [comment]
Hi @nedjwab 
Would love to follow you and have you follow me
What do you say?
here is my GitHub https://github.com/kayjeee
And my LinkedIn https://www.linkedin.com/in/kagiso-sebogodi/

**Review by mukhammadrizooff — COMMENTED** [review]
Hi @kayjeee, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that While you did well in your project milestone, I unfortunately cannot proceed with the review. Your GitHub workflows are currently not running which might be caused by the merge conflicts that you currently have. KIndly first resolve the merge conflits which will allow the GitHub workflows to run and then your project could be reviewed. Have a look at [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-on-github) resource to help you to resolve the merge conflicts🙏🏼 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,12 +1,13 @@
 class UsersController < ApplicationController
+  before_action :authenticate_user!, only: [:show]
   before_action :find_user, only: [:show]
```
Comment (BregornOriginal) [inline app/controllers/users_controller.rb:4]: - [ ] In your `posts_controller.rb` kindly add to the top of your class `PostsController` the method `load_and_authorize_resource` from `CanCanCan` to manage authorizations.

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new # Guest user
+
+    if user.role == 'admin'
+      can :manage, :all
+    else
+      can :read, :all
+      can :delete, [Post, Comment], author_id: user.id
```
Comment (BregornOriginal) [inline app/models/ability.rb:11]: - [ ] Excellent work implementing authorizations for a `user`. Kindly add the conditions to `manage` a `Comment` that belongs to a `user`.

Comment (BregornOriginal) [inline app/models/ability.rb:11]: - [ ] Remember that a user can `manage` a `post` which belongs to him. This means that can do CRUD abilities. Kindly add the conditions to `manage` a `Post` that belongs to a `user`.

**Review by BregornOriginal — CHANGES_REQUESTED** [review]
Hi @Team 

Good job so far!

There are some issues that you still need to work on to go to the next project but you are almost there!

giphy (5)

## Highlights

✅ Linters are passing
✅ Excellent work adding roles to the `user` 💯 
✅ README looks professional

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by IngDavidTM — APPROVED** [review]
Hi @kayjeee 🙋‍♂️, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations!  🎉💯🌟 

## Top highlight 🎯

- The Pull Request follows the GitFlow ✅ 
- The PR title and description are professional and descriptive ✅ 
- Linters are working properly ✅ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**